src/main/java/javax/jmdns/impl/DNSEntry.java

```diff
@@ -45,12 +45,16 @@ public abstract class DNSEntry {
         _recordType = type;
         _dnsClass = recordClass;
         _unique = unique;
-        _qualifiedNameMap = ServiceTypeDecoder.decodeQualifiedNameMapForType(this.getName());
+        _qualifiedNameMap = ServiceTypeDecoder.decodeQualifiedNameMapForType(this.getName(), type);
         String domain = _qualifiedNameMap.get(Fields.Domain);
         String protocol = _qualifiedNameMap.get(Fields.Protocol);
         String application = _qualifiedNameMap.get(Fields.Application);
         String instance = _qualifiedNameMap.get(Fields.Instance).toLowerCase();
-        _type = (application.length() > 0 ? "_" + application + "." : "") + (protocol.length() > 0 ? "_" + protocol + "." : "") + domain + ".";
+        String _type = (application.length() > 0 ? "_" + application + "." : "") + (protocol.length() > 0 ? "_" + protocol + "." : "") + (domain.length() > 0 ? domain + "." : "");
```

Comment:
Firstly, I'm missing unittests for this change. 
Next to that, without Null-checks, the `length()` operations are dangerous and will cause NPE's at some point. 

Since the extra Null-checks will cause this piece of code to grow, you might consider extracting a method out of this. That will also benefit testability.

Reply:
done! thanks